muspinsim/cpp/parallel.cpp

```diff
@@ -0,0 +1,239 @@
+#include "base.h"
+
+/*
+ * Computes V^\dagger (M \otimes 1_{d}) V between a complex vector V and a
+ * matrix M where 1_{d} an identity matrix of size d
+ *
+ * For speed, no checks are performed here. The matrix M should be Hermitian
+ * with size N and the shape of V should be (N*d, 1).
+ *
+ * @param V The vector
+ * @param M The matrix
+ * @param d The dimension of the identity matrix
+ *
+ * @return The result of the combined product. Will be a single value.
+ */
+double parallel::fast_measure(np_array_complex_t& V, np_array_complex_t& M, long int d) {
+    py::buffer_info V_info = V.request();
+    py::buffer_info M_info = M.request();
+
+    return parallel::fast_measure_ptr(
+        static_cast<std::complex<double>*>(V_info.ptr),
+        V_info.shape[0],
+        static_cast<std::complex<double>*>(M_info.ptr),
+        M_info.shape[0],
+        d);
+}
+
+/* Same as above, but only for use in C++ - uses pointers instead of numpy
```

Comment:
I'm conflicted on this, but I think I'd prefer it if we also included:
```
 * Computes V^\dagger (M \otimes 1_{d}) V between a complex vector V and a
 * matrix M where 1_{d} an identity matrix of size d
 *
 * For speed, no checks are performed here. The matrix M should be Hermitian
 * with size N and the shape of V should be (N*d, 1).
 ```
As part of the docstring (and similarly for the other "Python"/pointer methods). We're already duplicating other code comments such as 
```
// M[i, j] -> M[i * operator_dim + j] as column major
```
across the pointer functions, and I found those really useful when going through the code. But because I read `celio.cpp` first, I just started reading the (pointer) functions it called and so missed the useful content in the "Python" docstrings like that the dimensionality of `V` is `(N*d, 1)`. On the other hand it's duplication, and I should've just read all the docstrings beforehand.

It's also depedent on whether we still keep the Python functions, (see other comment above - the irony of this isn't lost on me).

Reply:
Yeah that's fine, if its ever changed its only in two places and can be copied, in this case I think its more important its clear for anyone looking at it in the future.